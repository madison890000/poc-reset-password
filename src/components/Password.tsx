import {ChangeEvent, useState} from "react";
import validatorPassword from "./ValidatePassWordReg";

interface IValidateResult {
    success: boolean;
    message?: string;
}

interface PasswordProps {
    validator?: RegExp;
    onValidate?: (result: IValidateResult) => void;
}

enum ValidateStatus {
    INIT = 'INIT',
    SUCCESS = 'SUCCESS',
    FAIL = 'FAIL',
}

const Password = ({onValidate}: PasswordProps) => {
    const [validateStatus, setValidateStatus] = useState<string>(ValidateStatus.INIT);
    const [message, setMessage] = useState<string>('');
    const [success, setSuccess] = useState<boolean>(false);
    const onChange = (e: ChangeEvent<HTMLInputElement>) => {
        const {success, message} = validatorPassword(e?.target?.value);
        setMessage(message ?? '');
        setSuccess(success);
        setValidateStatus(success ? ValidateStatus.SUCCESS : ValidateStatus.FAIL);
        onValidate?.({
            success,
            message
        })
    }
    const passTheValidation = validateStatus != ValidateStatus.INIT && success;
    return (
        <div style={{
            display: 'block'
        }}>
            <h3>
                Reset Password
            </h3>
            <div>
                <input
                    onChange={onChange}
                    data-testid="reset-password"
                    placeholder="Please input your password"
                    style={{
                        width: 200,
                        height: 30,
                    }}
                    type="password"
                />
            </div>
            {
                passTheValidation ? <div>password is fit the rules!</div> : (
                    <div style={{
                        color: 'red',
                        maxWidth: 200,
                    }}>
                        {message}
                    </div>
                )
            }
        </div>
    )
}

export default Password