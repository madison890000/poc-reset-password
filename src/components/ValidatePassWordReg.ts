export const VALIDATE_PASSWORD_REG = /^(?=.*\d)(?=(.*[!@#$%^&*()_+={}[\]:;"'<>,.?\/|\\~`-]){2}).{8,}$/;

export const LESS_THAN_8_CHARACTERS = 'password are at least 8 characters!';
export const SHOULD_INCLUDE_ONE_NUMBER_AND_TWO_SPECIAL_CHARACTERS = 'password should include one number and two special characters!';
const validatorPassword = (inputValue: string) => {
    if (inputValue.length < 8) {
        return {
            success: false,
            message: LESS_THAN_8_CHARACTERS
        }
    } else if (inputValue.length >= 8 && inputValue.length <= 15) {
        const pass = VALIDATE_PASSWORD_REG.test(inputValue);
        return {
            success: pass,
            message: pass ? '' : SHOULD_INCLUDE_ONE_NUMBER_AND_TWO_SPECIAL_CHARACTERS
        }
    } else {
        return {
            success: true,
        }
    }
}
export default validatorPassword