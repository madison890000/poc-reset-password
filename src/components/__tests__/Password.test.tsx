import React from 'react';
import Password from '../Password';
import renderer from 'react-test-renderer';
import {fireEvent, render, screen, waitFor} from '@testing-library/react';
import {LESS_THAN_8_CHARACTERS, SHOULD_INCLUDE_ONE_NUMBER_AND_TWO_SPECIAL_CHARACTERS} from "../ValidatePassWordReg";

describe('renders Password Component initial', () => {
    test('snapshot for Password Component initial', () => {
        const resetPasswordModule = renderer.create(<Password/>).toJSON();
        expect(resetPasswordModule).toMatchSnapshot();
    });
    test('Password Component should have a input', () => {
        render(<Password/>);
        const passwordInput = screen.getByTestId('reset-password');
        expect(passwordInput).toBeVisible();
    });
})

describe('Password Component change actions', () => {
    test('message:LESS_THAN_8_CHARACTERS when less than 8 characters', async () => {
        render(<Password/>);
        const passwordInput = screen.getByTestId('reset-password');
        fireEvent.change(passwordInput, {target: {value: 'abc123'}});
        await waitFor(() => {
            expect(screen.getByText(LESS_THAN_8_CHARACTERS)).toBeInTheDocument();
        })
    })

    test('message:LESS_THAN_8_CHARACTERS when less than 8 characters', async () => {
        render(<Password/>);
        const passwordInput = screen.getByTestId('reset-password');
        fireEvent.change(passwordInput, {target: {value: 'abc123abcsdd'}});
        await waitFor(() => {
            expect(screen.getByText(SHOULD_INCLUDE_ONE_NUMBER_AND_TWO_SPECIAL_CHARACTERS)).toBeInTheDocument();
        })
    })
    test('success when input correctly', async () => {
        render(<Password/>);
        const passwordInput = screen.getByTestId('reset-password');
        fireEvent.change(passwordInput, {target: {value: 'abc123abc__'}});
        await waitFor(() => {
            expect(screen.queryByText(SHOULD_INCLUDE_ONE_NUMBER_AND_TWO_SPECIAL_CHARACTERS)).toEqual(null);
        })
    })
})