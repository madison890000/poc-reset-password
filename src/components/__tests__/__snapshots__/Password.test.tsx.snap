// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Password Component initial snapshot for Password Component initial 1`] = `
<div
  style={
    Object {
      "display": "block",
    }
  }
>
  <h3>
    Reset Password
  </h3>
  <div>
    <input
      data-testid="reset-password"
      onChange={[Function]}
      placeholder="Please input your password"
      style={
        Object {
          "height": 30,
          "width": 200,
        }
      }
    />
  </div>
  <div
    style={
      Object {
        "color": "red",
        "maxWidth": 200,
      }
    }
  />
</div>
`;
